src/client/ws/client.rs

```diff
@@ -257,164 +255,228 @@ impl<Notif> Drop for Subscription<Notif> {
-					None => {
-						log::error!("Received notification on unknown subscription: {:?}", request_id);
-						continue;
-					}
-				};
-
-				match notifs_tx.try_send(result) {
-					Ok(()) => (),
-					// Channel is either full or disconnected, close it.
-					Err(e) => {
-						log::error!("Subscription ID: {:?} failed: {:?}", request_id, e);
-						let (_, unsubscribe) =
-							active_subscriptions.remove(&request_id).expect("Request is active checked above; qed");
-						close_subscription(&mut client, request_id, unsubscribe).await;
+				Some(Ok(jsonrpc::Response::Notif(notif))) => {
+					let sub_id = notif.params.subscription;
+					let request_id = match manager.get_request_id_by_subscription_id(&sub_id) {
+						Some(r) => r,
+						None => {
+							log::error!("Subscription ID: {:?} not found", sub_id);
+							continue;
+						}
+					};
+
+					match manager.as_active_subscription_mut(&request_id) {
+						Some(send_back_sink) => {
+							if let Err(e) = send_back_sink.try_send(notif.params.result) {
+								log::error!("Dropping subscription {:?} error: {:?}", sub_id, e);
+								manager.remove_active_subscription(request_id, sub_id).expect("subscription is active; checked above");
+							}
+						}
+						None => {
+							log::error!("Subscription ID: {:?} not an active subscription", sub_id);
+						},
 					}
 				}
-			}
-
-			// Request for the server to unsubscribe to us has succeeded.
-			Either::Right(Ok(RawClientEvent::Unsubscribed { request_id: _ })) => {}
-
-			Either::Right(Err(RawClientError::Inner(WsConnectError::Ws(SokettoError::UnexpectedOpCode(e))))) => {
-				log::error!(
-					"Client Error: {:?}, <https://github.com/paritytech/jsonrpsee/issues/154>",
-					SokettoError::UnexpectedOpCode(e)
-				);
-			}
-			Either::Right(Err(e)) => {
-				// TODO: https://github.com/paritytech/jsonrpsee/issues/67
-				log::error!("Client Error: {:?} terminating connection", e);
-				break;
-			}
+				Some(Err(e)) => {
+					log::error!("Error: {:?} terminating client", e);
+					return;
+				}
+			},
 		}
 	}
 }
 
-/// Close subscription in RawClient helper.
-/// Logs if the subscription couldn't be found.
-async fn close_subscription(client: &mut RawClient, request_id: RawClientRequestId, unsubscribe_method: String) {
-	match client.subscription_by_id(request_id).and_then(|s| s.into_active()) {
-		Some(mut sub) => {
-			if let Err(e) = sub.close(&unsubscribe_method).await {
-				log::error!("RequestID : {:?}, unsubscribe to {} failed: {:?}", request_id, unsubscribe_method, e);
+/// Process a response from the server.
+///
+/// Returns `Ok(_)` if the response was successful or if the error could be handled.
+/// Returns `Err(_)` if the response couldn't be handled.
+fn process_response(
+	manager: &mut RequestManager,
+	response: jsonrpc::Output,
+	subscription_capacity: usize,
+) -> Result<Option<(String, jsonrpc::Params)>, Error> {
+	let response_id = match response.id().as_number() {
+		Some(n) => *n,
```

Comment:
Could this be `let send_back_oneshot =  manager.complete_pending_method_call(response_id).ok_or_else(|| { Error::InvalidRequestId })?;`